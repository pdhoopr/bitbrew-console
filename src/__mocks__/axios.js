function getDefaults({ baseURL = '', headers = {} } = {}) {
  return {
    baseURL,
    headers: {
      common: headers,
    },
  };
}

function mockHttpMethod() {
  return jest.fn().mockResolvedValue({
    data: {},
  });
}

class AxiosMock {
  defaults = getDefaults();
  get = mockHttpMethod();
  post = mockHttpMethod();
  put = mockHttpMethod();
  delete = mockHttpMethod();
}

class ApiMock {
  api = new AxiosMock();

  create = jest.fn((config) => {
    this.api.defaults = getDefaults(config);
    return this.api;
  });

  reset = () => {
    this.api = new AxiosMock();
  };
}

export default new ApiMock();
