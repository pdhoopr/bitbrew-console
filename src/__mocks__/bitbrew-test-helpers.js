import diff from 'jest-diff';
import ShallowRenderer from 'react-test-renderer/shallow';

export function diffComponents(baseComponent, compareComponent) {
  const changes = diff(baseComponent, compareComponent, {
    aAnnotation: `<BaseComponent />`,
    bAnnotation: `<CompareComponent />`,
    contextLines: 3,
    expand: false,
  });
  return `Diff:\n${changes}`;
}

export function shallow(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

export const fixtures = {
  accessToken:
    'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE1MjUxMTUzNTAsImV4cCI6MTU1NjY1MTM1MCwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSJ9.4O-W8aeOOUFU8fkGTAFaAu9D9nK_l61yPomNG1Qkny8',
  orgs: {
    danlaw: {
      id: 'bd006616-3bd2-4783-a84d-b0fa4ed81135',
      properName: 'Danlaw',
    },
    azuga: {
      id: '8d9ec637-f3bb-4c46-83ee-11764d09f0b2',
      properName: 'Azuga',
    },
    bitbrew: {
      id: '94d9b0ca-a758-4502-8dcd-41d883e38714',
      properName: 'BitBrew',
    },
  },
  users: {
    bill: {
      email: 'bill@bitbrew.com',
      name: 'Bill Monaghan',
    },
    cynthia: {
      email: 'cynthia@bitbrew.com',
      name: 'Cynthia Hamilton',
    },
    alex: {
      email: 'bill@bitbrew.com',
      name: 'Alex Li',
    },
  },
};
