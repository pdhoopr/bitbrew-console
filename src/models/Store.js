import { autorun } from 'mobx';
import { flow, types } from 'mobx-state-tree';
import api from '../utils/api';
import Org from './Org';

export default types
  .model('Store', {
    orgs: types.optional(types.array(Org), []),
    storageKey: types.optional(types.string, 'bitbrew-console'),
    token: types.maybe(types.string),
  })
  .views((self) => ({
    get isSignedIn() {
      return !!self.token;
    },
  }))
  .actions((self) => ({
    afterCreate() {
      self.token = localStorage.getItem(self.storageKey);
    },
    signIn({ accessToken }) {
      self.token = accessToken;
      localStorage.setItem(self.storageKey, accessToken);
    },
    signOut() {
      self.token = null;
      localStorage.removeItem(self.storageKey);
    },
    configureApi: autorun(() => {
      api.configure({
        token: self.token,
      });
    }),
    listOrgs: flow(function* listOrgs() {
      const { items } = yield api.orgs.list();
      self.orgs = items
        .map((org) => Org.create(org))
        .sort((a, b) =>
          a.properName.toLowerCase().localeCompare(b.properName.toLowerCase()),
        );
    }),
  }));
