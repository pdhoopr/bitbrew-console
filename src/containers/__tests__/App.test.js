import axios from 'axios';
import { diffComponents, fixtures } from 'bitbrew-test-helpers';
import { shallow } from 'enzyme';
import { Provider } from 'mobx-react';
import React from 'react';
import { Simulate, render } from 'react-testing-library';
import Store from '../../models/Store';
import ConnectedApp from '../App';

const App = ConnectedApp.wrappedComponent;

beforeEach(() => {
  localStorage.clear();
  axios.reset();
});

test('renders DOM for the app when signed out', () => {
  const dom = shallow(<App isSignedIn={false} />);
  expect(dom).toMatchSnapshot();
});

test('renders different DOM for the app when signed in', () => {
  const base = <App isSignedIn={false} />;
  const comp = <App isSignedIn />;
  const diff = diffComponents(base, comp);
  expect(diff).toMatchSnapshot();
});

test('allows signing in and out of the app', async () => {
  axios.api.get.mockResolvedValueOnce({
    data: {
      items: [],
    },
  });
  const { getByLabelText, getByTestId, getByText } = render(
    <Provider store={Store.create()}>
      <ConnectedApp />
    </Provider>,
  );
  expect(getByText('Sign in')).toBeDefined();
  const accessToken = getByLabelText('Access Token');
  accessToken.value = fixtures.accessToken;
  Simulate.change(accessToken);
  Simulate.submit(getByTestId('sign-in-form'));
  expect(getByText('Welcome!')).toBeDefined();
  const signOut = getByText('Sign out');
  expect(signOut).toBeDefined();
  Simulate.click(signOut);
  expect(getByText('Sign in')).toBeDefined();
});
