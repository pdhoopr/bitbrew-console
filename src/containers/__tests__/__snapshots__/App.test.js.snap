// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DOM for the app when signed out 1`] = `<inject-SignInPage />`;

exports[`renders different DOM for the app when signed in 1`] = `
Diff:
- <BaseComponent />
+ <CompareComponent />

- <inject-SignInPage />
+ <inject-ListOrgsPage />
`;
